janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java

```diff
@@ -103,45 +109,140 @@ public static JanusGraphTransaction getTx(Traversal.Admin<?, ?> traversal) {
-
-                if (parentStep instanceof RepeatStep && multiQueryCompatibleSteps.contains(parentStep)) {
-                    RepeatStep repeatStep = (RepeatStep)parentStep;
-                    List<RepeatEndStep> repeatEndSteps = TraversalHelper.getStepsOfClass(RepeatEndStep.class, repeatStep.getRepeatTraversal());
-                    if (repeatEndSteps.size() == 1) {
-                        // Want the RepeatEndStep so the start of one iteration can feed into the next
-                        multiQueryCompatibleSteps.remove(parentStep);
-                        multiQueryCompatibleSteps.add(repeatEndSteps.get(0));
-                    }
-                }
+    public static List<Step> getAllMultiQueryPositionsForMultiQueriable(Step<?, ?> multiQueriable) {
+        List<Step> multiQueryStepLocations = new ArrayList<>();
+        Queue<Step> rawLocations = new LinkedList<>();
+        Step currentStep = multiQueriable;
+
+        do {
+            rawLocations.add(currentStep);
+            currentStep = currentStep.getTraversal().getParent().asStep();
+        } while (currentStep instanceof RepeatStep);
+
+        while (!rawLocations.isEmpty()) {
+            currentStep = rawLocations.poll();
+            Optional<Step> positionInLocalTraversal = getLocalMultiQueryPositionForStep(currentStep);
+            if (positionInLocalTraversal.isPresent()) {
+                multiQueryStepLocations.add(positionInLocalTraversal.get());
+            } else {
+                rawLocations.add(currentStep.getTraversal().getParent().asStep());
             }
         }
-        return new ArrayList<>(multiQueryCompatibleSteps);
+
+        return multiQueryStepLocations;
+    }
+
+    /**
+     * For a MultiQuery compatible step, this method searches the correct position in the step's traversal at which
+     * a <code>JanusGraphMultiQueryStep</code> should be inserted. Only the traversal of the given step is considered,
+     * parent and child traversals are not taken into account.
+     * @param step The MultiQuery compatible step.
+     * @return The step before which a <code>JanusGraphMultiQueryStep</code> should be inserted.
+     * @see org.janusgraph.graphdb.tinkerpop.optimize.step.JanusGraphMultiQueryStep
+     */
+    public static Optional<Step> getLocalMultiQueryPositionForStep(Step<?, ?> step) {
+        Step currentStep = step;
+        Step previousStep = step.getPreviousStep();
+        while ((previousStep instanceof SideEffectStep || previousStep instanceof ProfileStep)
+            && !(previousStep instanceof InjectStep)) {
+            currentStep = previousStep;
+            previousStep = previousStep.getPreviousStep();
+        }
+        if (previousStep.equals(EmptyStep.instance()) || previousStep instanceof StartStep) {
+            final Step parentStep = step.getTraversal().getParent().asStep();
+            if (isMultiQueryCompatibleParent(parentStep) && !(parentStep instanceof RepeatStep)) {
```

Comment:
I think it's better to put `!(parentStep instanceof RepeatStep)` before `isMultiQueryCompatibleParent(parentStep)` because `isMultiQueryCompatibleParent` probably will be slower to check